src/Google/GlobalSiteTag.php

```diff
@@ -156,14 +195,241 @@ public function maybe_display_event_snippet( string $ads_conversion_id, string $
+	 */
+	public function display_purchase_page_snippet(): void {
+		// Only display on the order confirmation page.
+		if ( ! is_order_received_page() ) {
+			return;
+		}
+		$order_id = $this->wp->get_query_vars( 'order-received', 0 );
+		if ( empty( $order_id ) ) {
+			return;
+		}
+		$order = wc_get_order( $order_id );
+
+		$item_info = '';
+		foreach ( $order->get_items() as $item_id => $item ) {
+			$product_id   = $item->get_product_id();
+			$product_name = $item->get_name();
+			$quantity     = $item->get_quantity();
+			$price        = $item->get_subtotal();
+			$item_info    = $item_info . sprintf(
+				'{
+                    "id": "gla_%s",
+                    "price": %s,
+                    "google_business_vertical": "retail",
+                    "name":"%s",
+                    "quanitity":"%s",
+                }',
+				esc_js( $product_id ),
+				esc_js( $price ),
+				esc_js( $product_name ),
+				esc_js( $quantity ),
+			);
+
+		}
+
+		$is_new_customer = false;
+		if ( $order->get_user_id() ) {
+			$total_orders = wc_get_customer_order_count( $order->get_user_id() );
+		} else {
+			$total_orders = WC_Order_Export_Data_Extractor::get_customer_order_count_by_email( $order->get_billing_email() );
+		}
+		$is_new_customer = ( $total_orders === 1 ) ? 'true' : 'false';
+		$language        = $this->wp->get_locale();
+		if ( 'en_US' === $language ) {
+			$language = 'English';
+		}
+		printf(
+			'<script>
+            gtag("event", "purchase",
+				{
+                    "developer_id.%s": "true",
+                    "ecomm_pagetype": "purchase",
+                    "send_to": "GLA",
+                    transaction_id": "%s",
+                    "currency": "%s",
+                    "country": "%s",
+                    "value": "%s",
+                    "new_customer": "%s",
+                    "tax": "%s",
+                    "shipping": "%s",
+                    "delivery_posatal_code": "%s",
+                    "aw_feed_country": "%s",   
+                    "aw_feed_language": "%s",                 
+                    items: [' . esc_html( $item_info ) . ']}); </script>',
+			esc_js( self::DEVELOPER_ID ),
+			esc_js( $order->get_id() ),
+			esc_js( $order->get_currency() ),
+			esc_js( WC()->countries->get_base_country() ),
+			esc_js( $order->get_total() ),
+			esc_js( $is_new_customer ),
+			esc_js( $order->get_cart_tax() ),
+			esc_js( $order->get_total_shipping() ),
+			esc_js( $order->get_shipping_postcode() ),
+			esc_js( WC()->countries->get_base_country() ),
+			esc_js( $language ),
+		);
+	}
+
+	/**
+	 * Display the JavaScript code to track the add to cart button.
+	 *
+	 * @param string $message Add to cart messages.
+	 * @param array  $products Product ID list.
+	 */
+	public function custom_action_add_to_cart( $message, $products ) {
+		// Only display this tag info after click the add to cart button .
+		$product = wc_get_product( array_key_first( $products ) );
```

Comment:
If this filter supports multiple items, why do we only get the first item? The add_to_cart event supports an array of items.

Reply:
In my understanding, customer could add single product or multiple same products with clicking the "add to cart" button on one product page of current template. We could get all the required data from one single product. Therefore I only read the first product info here. Please correct me if I'm wrong.

Reply:
In most cases that is correct. A [grouped product](https://github.com/woocommerce/woocommerce/blob/6.3.1/plugins/woocommerce/includes/class-wc-form-handler.php#L838) is an exception though, it will pass an array of products to that function.

Although I'm not sure how relevant this is if the solution is not compatible with blocks. Have you done some testing to see if this code is triggered when using the All Products block?

Reply:
It makes sense. Thanks. Updated.